File: BPM-Detection(Python)/Functions/dynamicRangeCompression.py
import numpy as np

def dynamicRangeCompression(x, compressionThreshold, compressionRatio):
    """
    动态范围压缩函数

    参数：
        x                   : 输入音频信号（1D numpy array）
        compressionThreshold: 压缩阈值
        compressionRatio    : 压缩比

    返回：
        x_out : 压缩后的信号
    """
    x_out = np.copy(x)
    x_abs = np.abs(x)

    # 找出超过阈值的部分
    aboveThreshold = x_abs > compressionThreshold

    # 对超过阈值的部分进行压缩
    x_out[aboveThreshold] = np.sign(x[aboveThreshold]) * (
        compressionThreshold + (x_abs[aboveThreshold] - compressionThreshold) / compressionRatio
    )

    return x_out

File: Functions/dynamicRangeCompression.m
function x_out = dynamicRangeCompression(x, compressionThreshold, compressionRatio)
%Initialize output signal
x_out = x;
x_abs = abs(x);
%Find the part that needs to be compressed (the part that exceeds the threshold)
aboveThreshold = x_abs > compressionThreshold;

%Apply compression to the parts that exceed the threshold, only changing the parts that need to be compressed
x_out(aboveThreshold) = sign(x(aboveThreshold)) .* ...
     (compressionThreshold + (x_abs(aboveThreshold) - compressionThreshold) / compressionRatio);

end
